### PR TITLE
[feat] Attemp to find a better way to seriable preprocess

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR attempts to add a better serialisation mechanism to Lightning Flash.

Fixes # (issue)

## Before submitting
- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/lightning-flash/tree/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests? [not needed for typos/docs]
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/lightning-flash/blob/master/CHANGELOG.md)?

## PR review
 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: flash/data/process.py [hunk 1]
```diff
@@ -286,25 +304,92 @@ def __init__(
...
+            try:
+                gitdiff = subprocess.check_output(["git", "diff"]).decode()
+            except:
+                gitdiff = ""
+            state_dict = state_dict_fn(*args, **kwargs)
+            state_dict["_meta"] = {}
+            state_dict["_meta"]["module"] = self.__module__
+            state_dict["_meta"]["class_name"] = self.__class__.__name__
+            state_dict["_meta"]["args"] = self._args
+            state_dict["_meta"]["kwargs"] = self._kwargs
```
Comment (ethanwharris) [inline flash/data/process.py:356]: Should also add a `state_dict["_meta"]["_state"] = self._state` in order to preserve the labels (see here: https://github.com/PyTorchLightning/lightning-flash/blob/master/flash/data/process.py#L52)

File: flash/data/process.py [hunk 2]
```diff
@@ -286,25 +304,92 @@ def __init__(
...
+        pass
+
+    @classmethod
+    def from_state_dict(cls, state_dict: Dict[str, Any]):
+        if cls == Preprocess and "_meta" in state_dict:
+            try:
+                import importlib
+                meta = state_dict["_meta"]
+                _cls = getattr(importlib.import_module(meta["module"]), meta["class_name"])
+                return _cls(*meta["args"], **meta["kwargs"])
```
Comment (ethanwharris) [inline flash/data/process.py:387]: And here should do `object._state = meta["_state"]` to reload it.

File: flash/data/process.py [hunk 3]
```diff
@@ -128,7 +133,31 @@ def validating(self, val: bool) -> None:
...
+class AbstractPreprocess(ABC):
+
+    @abstractstaticmethod
+    def version() -> str:
+        """
+        Override to versioned to version your preprocess.
+        """
+        pass
+
+    @abstractmethod
```
Comment (carmocca) [inline flash/data/process.py:145]: Why not rely on the `nn.Module` `state_dict`/`load_state_dict` for this?

Preprocess does inherit it

File: flash/data/process.py [hunk 4]
```diff
@@ -295,17 +324,34 @@ def __init__(
...
-    # todo (tchaton) Add a warning if a transform is provided, but the hook hasn't been overriden !
+    def _save_to_state_dict(self, destination, prefix, keep_vars):
+        preprocess_state_dict = self.save_state_dict()
+        if not isinstance(preprocess_state_dict, Dict):
+            raise MisconfigurationException("save_state_dict should return a dictionary")
+        preprocess_state_dict["_meta"] = {}
+        preprocess_state_dict["_meta"]["module"] = self.__module__
+        preprocess_state_dict["_meta"]["class_name"] = self.__class__.__name__
+        preprocess_state_dict["_meta"]["version"] = self.version()
+        try:
```
Comment (carmocca) [inline flash/data/process.py:347]: parsing and setting any git attributes should be part of version, for example from my project:

https://github.com/jpuigcerver/PyLaia/blob/1b2e864247f1bfb8d95ac1910de9c52df71c017a/laia/__init__.py#L21-L31

File: flash/data/process.py [hunk 5]
```diff
@@ -128,7 +133,31 @@ def validating(self, val: bool) -> None:
...
 
-class Preprocess(Properties, Module):
+class AbstractPreprocess(ABC):
+
+    @abstractstaticmethod
+    def version() -> str:
+        """
+        Override to versioned to version your preprocess.
+        """
+        pass
```
Comment (carmocca) [inline flash/data/process.py:143]: Why would the preprocess object have specific version? why not use `flash.__version__`?

File: flash/core/model.py [hunk 6]
```diff
@@ -451,3 +453,35 @@ def _instantiate_scheduler(self, optimizer: Optimizer) -> _LRScheduler:
...
+        self, state_dict, prefix, local_metadata, strict, missing_keys, unexpected_keys, error_msgs
+    ):
+        if 'preprocess.state_dict' in state_dict:
+            try:
+                preprocess_state_dict = state_dict["preprocess.state_dict"]
+                meta = preprocess_state_dict["_meta"]
+                module = meta["module"]
+                class_name = meta["class_name"]
+                checkpoint_version = meta["version"]
+                commit_sha = meta["commit_sha"]
```
Comment (carmocca) [inline flash/core/model.py:467]: is this just checking if an exception is raised? why not `assert hasattr(...)`?

File: flash_examples/predict/image_classification.py [hunk 7]
```diff
@@ -19,7 +19,7 @@
 download_data("https://pl-flash-data.s3.amazonaws.com/hymenoptera_data.zip", "data/")
 
 # 2. Load the model from a checkpoint
-model = ImageClassifier.load_from_checkpoint("https://flash-weights.s3.amazonaws.com/image_classification_model.pt")
+model = ImageClassifier.load_from_checkpoint("image_classification_model.pt")
```
Comment (ethanwharris) [inline flash_examples/predict/image_classification.py:22]:
```suggestion
model = ImageClassifier.load_from_checkpoint("https://flash-weights.s3.amazonaws.com/image_classification_model.pt")
```

File: flash/data/process.py (hunk 3 above)
Comment (tchaton) [reply flash/data/process.py:145]: Currently, state_dict from nn.Module calls Preprocess `_save_to_state_dict` nn.Module hook which calls the users `save_state_dict` hook which is mandatory as `abstractmethod`.

By doing so, users are force to properly serialise their Preprocess objects.

File: flash/data/process.py [hunk 8]
```diff
@@ -295,17 +324,36 @@ def __init__(
...
         self._test_collate_in_worker_from_transform: Optional[bool] = None
 
-        self.train_transform = convert_to_modules(self._check_transforms(train_transform, RunningStage.TRAINING))
-        self.val_transform = convert_to_modules(self._check_transforms(val_transform, RunningStage.VALIDATING))
-        self.test_transform = convert_to_modules(self._check_transforms(test_transform, RunningStage.TESTING))
-        self.predict_transform = convert_to_modules(self._check_transforms(predict_transform, RunningStage.PREDICTING))
+        self._train_transform = self._check_transforms(train_transform, RunningStage.TRAINING)
+        self._val_transform = self._check_transforms(val_transform, RunningStage.VALIDATING)
+        self._test_transform = self._check_transforms(test_transform, RunningStage.TESTING)
+        self._predict_transform = self._check_transforms(predict_transform, RunningStage.PREDICTING)
```
Comment (edgarriba) [inline flash/data/process.py:330]: apart from next 4 lines, are these variables used anywhere else ? if not, maybe no need to be members

File: flash/data/process.py (hunk 4 above)
Comment (tchaton) [reply flash/data/process.py:349]: I thought about it, but I would prefer to leave this responsibility on the users. 
They should dump the version if the internal behaviour changed.
I am just saving the commit_sha so they can found this commit back on their history potentially.

File: flash/data/process.py (hunk 5 above)
Comment (tchaton) [reply flash/data/process.py:143]: Because new way of processing the data might be implemented and the version should be dump.

File: flash/text/classification/data.py [hunk 9]
```diff
@@ -82,6 +83,24 @@ def __init__(
             class_to_label_mapping[cls] = label
         self.set_state(ClassificationState(class_to_label_mapping))
 
+    @staticmethod
+    def version() -> str:
+        return "0.0.1"
+
+    def save_state_dict(self) -> Dict[str, Any]:
```
Comment (edgarriba) [inline flash/text/classification/data.py:90]: this is not saving, right ? `get_state_dict` ?

File: flash/text/classification/data.py [hunk 10]
```diff
@@ -82,6 +83,24 @@ def __init__(
...
+            "input": self.input,
+            "backbone": self.backbone,
+            "max_length": self.max_length,
+            "target": self.target,
+            "filetype": self.filetype,
+            "label_to_class_mapping": self.label_to_class_mapping,
+        }
+
+    @classmethod
+    def load_state_dict(cls, state_dict: Dict[str, Any], keep_vars: bool):
```
Comment (edgarriba) [inline flash/text/classification/data.py:101]: `keep_vars` not used

File: flash/core/model.py [hunk 11]
```diff
@@ -451,3 +453,39 @@ def _instantiate_scheduler(self, optimizer: Optimizer) -> _LRScheduler:
...
+                _data_pipeline_state = meta["_data_pipeline_state"]
+                cls = getattr(import_module(meta["module"]), meta["class_name"])
+                del preprocess_state_dict["_meta"]
+                current_version = cls.version()
+                if current_version != checkpoint_version:
+                    rank_zero_warn(
+                        f"The preprocess `{module}.{class_name}` was generated with a different version "
+                        f"current: {current_version} checkpoint: {checkpoint_version}. "
+                        "The behaviour might not be the one expected."
+                        f"Hint: Check commit: {commit_sha}.", UserWarning
```
[the remark this answers is not in the document]
Comment (tchaton) [reply flash/core/model.py:478]: Done !

File: flash/data/process.py [hunk 12]
```diff
@@ -128,7 +133,31 @@ def validating(self, val: bool) -> None:
             self._running_stage = None
 
 
-class Preprocess(Properties, Module):
+class AbstractPreprocess(ABC):
+
+    @abstractstaticmethod
+    def version() -> str:
+        """
+        Override to versioned to version your preprocess.
```
[the remark this answers is not in the document]
Comment (tchaton) [reply flash/data/process.py:141]: Done !

File: flash_examples/predict/image_classification.py (hunk 7 above)
Comment (tchaton) [reply flash_examples/predict/image_classification.py:22]: Done !

File: flash/data/process.py (hunk 8 above)
Comment (tchaton) [reply flash/data/process.py:330]: Done !

File: flash/text/classification/data.py (hunk 9 above)
Comment (tchaton) [reply flash/text/classification/data.py:90]: Done !

File: flash/data/process.py [hunk 13]
```diff
@@ -295,17 +317,35 @@ def __init__(
...
+            raise MisconfigurationException("get_state_dict should return a dictionary")
+        preprocess_state_dict["_meta"] = {}
+        preprocess_state_dict["_meta"]["module"] = self.__module__
+        preprocess_state_dict["_meta"]["class_name"] = self.__class__.__name__
+        preprocess_state_dict["_meta"]["_state"] = self._state
+        try:
+            commit_sha = subprocess.check_output(["git", "rev-parse", "HEAD"]).decode("ascii").strip()
+        except subprocess.CalledProcessError:
+            commit_sha = "n/a"
+        preprocess_state_dict["_meta"]["commit_sha"] = commit_sha
```
Comment (ethanwharris) [inline flash/data/process.py:345]: This is never used, do we still need it?

**Review by ethanwharris — APPROVED** [review]
LGTM. Just a query about commit sha

File: flash/data/process.py (hunk 13 above)
Comment (tchaton) [reply flash/data/process.py]: Removed !
